### PR TITLE
Dblatcher/apps email sign up placement 2

### PR DESCRIPTION
**Pull request by dblatcher**
## What does this change?
[no visible changes]
Defines a function to insert a `NewsletterSignup` element into an `Item.body` in apps rendering, if the article has a `promotedNewsletter` and is "Standard" (IE does not a blog, interactive page etc). The function is not called as the component to be rendered is no functional yet.

 The exported insertNewsletterIntoItem function will try to insert a `NewsletterSignUp` according to the following rules:
- Don't put straight after bold text
- Don't put under headings
- Must have plain body text above and below
- Must be within 3 elements from the middle of the article
- Should not be next to an adSlot
- The best place is the last place meeting the criteria above.

If no suitable place is found, the article body is not modified.

## How to trial

In "apps-rendering/src/server/page.tsx", import the function and edit the start of the `render` as below.
```typescript
function render(
	imageSalt: string,
	request: RenderingRequest,
	getAssetLocation: (assetName: string) => string,
	page: Option<string>,
): Page {
	const _item = fromCapi({ docParser, salt: imageSalt })(request, page);
	_item.promotedNewsletter = {
		kind: 0,
		value: {
			identityName: 'patriarchy',
			description:
				'Reviewing the most important stories on feminism and sexism and those fighting for equality',
			name: 'The Week in Patriarchy',
			frequency: 'Weekly',
			theme: 'opinion',
			successDescription: 'signed up',
		},
	};
	const item = insertNewsletterIntoItem(_item);
```

## Why?
The new sign-up blocks are implemented on DCR and the equivalent feature is being rolled out on App-rendering.
When the `NewsletterSignup` is functional, the placement logic can be used to add the `NewsletterSignUp` BodyElement to the Item.

### CODE REVIEW

**Review by georgeblahblah — COMMENTED** [review]
Thanks for reworking this following the work completed in #5767!

I had some general comments around types. I also wondered if:

- we should think about removing the debugging statements and logging before merging?
- we should take another look at how to avoid Ad Slots. I have a feeling the current ad slot placement logic should avoid newsletter sign ups. I think that logic fits better in the `ads` module, rather than trying to predict an ad placement here

**Comment by dblatcher** [comment]
> * we should think about removing the debugging statements and logging before merging?

I'm in two minds about that - I don't like leaving debug code lying around, but also wouldn't want to re-create it every time the logic needed updating. Without the visualisation, it would be very hard to follow how the function was working. 
From our experience with DCR, I'm expecting the ruleset to change after user feedback.
If the debug logging function does survive in some form, it could probably be better separated from the core functions.

> * we should take another look at how to avoid Ad Slots. I have a feeling the current ad slot placement logic should avoid newsletter sign ups. I think that logic fits better in the `ads` module, rather than trying to predict an ad placement here

That makes sense - although I assume we'd want to avoid the Sign Ups reducing the total number of ads. I didn't feel entitled to change how ads are placed but the 'predict ads' approach is not  ideal. Maybe the best thing would be to remove that from this PR and look separately at how to stop Ad Slots and Sign Ups colliding?

It might be easier if the Ad Slots and Sign Ups where inserted  in the same step using a common set of rules. Its a similar operation.

File: apps-rendering/src/server/insertNewsletterIntoItem.ts [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+
+function insertNewsletterIntoBody(
+	item: Item & { body: Body },
+	newsletterSignUp: NewsletterSignUp,
+): Item {
+	const insertIndex = findInsertIndex(item.body);
+	if (insertIndex === -1) {
+		logger.warn(
+			`Unable to find suitable place for NewsletterSignUp: ${item.webUrl}`,
+		);
```
[the remark this answers is not in the document]
Comment (dblatcher) [reply apps-rendering/src/server/insertNewsletterIntoItem.ts:267]: makes sense, I'll change it to `info`

File: apps-rendering/src/server/insertNewsletterIntoItem.ts [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+		...item.body.slice(insertIndex),
+	];
+	return item;
+}
+
+function insertNewsletterIntoItem(item: Item): void {
+	const newsletterSignUp = buildBodyElement(item);
+	if (!newsletterSignUp) {
+		return;
+	}
```
[the remark this answers is not in the document]
Comment (dblatcher) [reply apps-rendering/src/server/insertNewsletterIntoItem.ts:282]: yes, that works.

File: apps-rendering/src/server/insertNewsletterIntoItem.ts [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+ * counting paragraph elements in ReactNodes, whereas here
+ * we are looking at bodyElement before they are have been
+ * rendered.
+ *
+ * @param contentOnlyBody an subset of an article Body, with the
+ * non-rendered elements (errors and whitespace) removed.
+ * @returns the indices of the elements that are expected to
+ * have an adslot placed before them.
+ */
+function predictIndicesOfParagraphsAfterAdslots(
```
[the remark this answers is not in the document]
Comment (dblatcher) [reply apps-rendering/src/server/insertNewsletterIntoItem.ts:101]: It is a hard requirement, but not once that necessarily needs to be implemented in this PR - the components aren't going live yet, so we could remove the logic and look separately at how to  address ads.

Currently, the ads are placed after the `BodyElements` have been rendered to `ReactNodes`. I don't think this kind of logic would work well at stage because we've then lost the structure from the data model.  Since [#5767](https://github.com/guardian/dotcom-rendering/pull/5767), there are `BodyElements` for individual paragraphs, so it might be moving the ad placement to be before the render to `ReactNodes` (and then have placing Sign Ups after placing ads, but both before the React stage).

File: apps-rendering/src/server/insertNewsletterIntoItem.ts [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+			const { doc } = result.value;
+			if (doc.textContent?.trim().length === 0) {
+				return ElementCategory.WhiteSpace;
+			}
+
+			if (doc.nodeName === 'P') {
+				if (doc.nodeType === 1) {
+					const element = doc as Element;
+					// for paragraphs containing only single <strong> child ("bold text"),
+					// the sign up block can be put before them, but not after them.
```
[the remark this answers is not in the document]
Comment (dblatcher) [reply apps-rendering/src/server/insertNewsletterIntoItem.ts:53]: Yes - If I'm understanding you correctly. This is to account for cases where the contributor uses a line of bold text in composer - we consider these to have been intended to act as headings and not allow a Sign Up after them.
Maybe `ElementCategory.PseudoHeading` would be a better name than `ElementCategory.BoldParagraphText`.

File: apps-rendering/src/server/insertNewsletterIntoItem.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (georgeblahblah) [reply apps-rendering/src/server/insertNewsletterIntoItem.ts:101]: yes, that makes sense. We had a discussion this morning about how to take this forward and think it's a good idea to:

- keep the newsletter signup placement logic here
- if necessary, teach ads how to avoid newsletter signups for now
- later, look into refactoring ad placement to also happen pre-render

File: apps-rendering/src/server/insertNewsletterIntoItem.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (georgeblahblah) [reply apps-rendering/src/server/insertNewsletterIntoItem.ts:53]: hmm yes. We don't currently parse H3s at the parsing step yet, although they are modeled in `BodyElement` (and you handle them above). I think the naming is fine (or you could use `HeadingThree` for now as well). We might want to look into parsing H3s another time

File: apps-rendering/src/server/insertNewsletterIntoItem.ts [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+	const { value: newsletter } = item.promotedNewsletter;
+	return {
+		kind: ElementKind.NewsletterSignUp,
+		...newsletter,
+		theme: stringToPillar(newsletter.theme),
+	};
+}
+
+function insertNewsletterIntoBody(
+	item: Item & { body: Body },
```
[the remark this answers is not in the document]
Comment (dblatcher) [reply apps-rendering/src/server/insertNewsletterIntoItem.ts]: I've refactored the main functions to use Results and avoid mutating objects. The error case is needed - part of the brief was to include logging so there's visibility on which / how many article can 't 'fit' a sign-up so we can check if the rules are too restrictive.

File: apps-rendering/src/server/insertNewsletterIntoItem.ts [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+		case ArticleDesign.Recipe:
+		case ArticleDesign.MatchReport:
+		case ArticleDesign.Interview:
+		case ArticleDesign.Editorial:
+		case ArticleDesign.Obituary:
+			insertNewsletterIntoBody(item, newsletterSignUp);
+			break;
+		default:
+			break;
+	}
```
[the remark this answers is not in the document]
Comment (dblatcher) [reply apps-rendering/src/server/insertNewsletterIntoItem.ts:309]: 👍

File: apps-rendering/src/server/insertNewsletterIntoItem.ts [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+ *	 - Must have plain body text above and below
+ *   - Must be within 3 elements from the middle of the article
+ *   - Should not be next to a adSlot
+ *   - The best place is the last place meeting the criteria above.
+ *
+ * @param body an Item.Body
+ * @returns the best index in that body to insert a sign-up block, or -1
+ * if there is no suitable place
+ */
+function findInsertIndex(body: Body): number {
```
[the remark this answers is not in the document]
Comment (dblatcher) [reply apps-rendering/src/server/insertNewsletterIntoItem.ts]: 👍

File: apps-rendering/src/server/insertNewsletterIntoItem.ts [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+import type { Body, BodyElement, NewsletterSignUp } from 'bodyElement';
+import { ElementKind } from 'bodyElementKind';
+import type { Item } from 'item';
+import { logger } from 'logger';
+import { Result } from 'result';
+import { stringToPillar } from 'themeStyles';
+
+// ----- Constants ----- //
+
+const enum ElementCategory {
```
[the remark this answers is not in the document]
Comment (dblatcher) [reply apps-rendering/src/server/insertNewsletterIntoItem.ts:15]: It wasn't a conscious decision -  I've not used `enums` so may have just found an example using the `const enum` first.

File: apps-rendering/src/server/insertNewsletterIntoItem.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (dblatcher) [reply apps-rendering/src/server/insertNewsletterIntoItem.ts]: Great - I've removed the 'predict ads' part. It would be good to pair on changes to the ads module.

**Comment by georgeblahblah** [comment]
> I don't like leaving debug code lying around, but also wouldn't want to re-create it every time the logic needed updating

I see what you're saying. I think the best we could do in this case would be to write a test suite to represent what we expect to happen for certain inputs. That way, we should know if something breaks that expectation without having to manually run different variations and compare debug output.

> Maybe the best thing would be to remove that from this PR and look separately at how to stop Ad Slots and Sign Ups colliding

Just as a summary: we looked at this, and determined that ads should already avoid newsletter signups. No additional implementation was required to achieve this, so we could safely remove the ad-related code from the newsletter placement module.

> It might be easier if the Ad Slots and Sign Ups where inserted in the same step using a common set of rules. Its a similar operation.

This is a really interesting point, and something we've talked about as a team. It would be great if this area could be simplified, and there's maybe something we can learn from the Commercial team's approach to building spacefinder!

File: apps-rendering/src/server/insertNewsletterIntoItem.ts [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+ *	 - Don't put under headings
+ *	 - Must have plain body text above and below
+ *   - Must be within 3 elements from the middle of the article
+ *   - Should not be next to an adSlot (NOT IMPLEMENTED HERE)
+ *   - The best place is the last place meeting the criteria above.
+ *
+ * @param body an Item.Body
+ * @returns A Result with the insertIndex or an error
+ */
+function findInsertIndex(body: Body): Result<string, number> {
```
Comment (georgeblahblah) [inline apps-rendering/src/server/insertNewsletterIntoItem.ts:98]: I had an idea for this function which might simplify things a little bit, but I'd be very interested to hear your thoughts.

I really like the idea of `getRange()`, and so I wanted to see if we could take even more advantage of that return value. Given `getRange()` gives us the first and last possible places to put the newsletter signup, I think we could narrow all processing between these two values.

In pseudocode, the algorithm then looks a bit like this:

```
const [firstPossibleIndex, lastPossibleIndex] = getRange(elements);

const getInsertionPoint(min: number, currentIndex: number, bodyElements: BodyElement[]): Result<string, number> {
  if (currentIndex <= min) return Result.err('could not find a place')
  if (isSuitable(currentIndex, bodyElements)) return Result.ok(currentIndex)
  return getInsertionPoint(min, currentIndex - 1, bodyElements)
}
```

where we start at the last possible point of insertion, and return early if it is correct. otherwise, keep stepping backwards until a place is found or nothing is found.

the definition of `isSuitable` is then a bit different, where it just checks to make sure the element _at_ `currentIndex` and the element _before_ `currentIndex` meet the criteria (not after a heading, not before a bold paragraph etc).

**Review by georgeblahblah — COMMENTED** [review]
Thanks for taking a look at the other comments! It's looking really good. 

I think there is still some unresolved decisions around debug logging and whether we could reduce the need for it with test cases.

Otherwise, I had an idea about the function which finds the place to insert the newsletter, and I'm interested to hear your thoughts!

Please reach out if it's easier to talk through any of the points on a call or in more detail

File: apps-rendering/src/server/insertNewsletterIntoItem.ts [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+}
+
+/**
+ *  Find the place to insert a insert a sign-up block in to an article body,
+ *  using these rules:
+ *	 - Don't put straight after bold text
+ *	 - Don't put under headings
+ *	 - Must have plain body text above and below
+ *   - Must be within 3 elements from the middle of the article
+ *   - Should not be next to an adSlot (NOT IMPLEMENTED HERE)
```
[the remark this answers is not in the document]
Comment (dblatcher) [reply apps-rendering/src/server/insertNewsletterIntoItem.ts:92]: > > I don't like leaving debug code lying around, but also wouldn't want to re-create it every time the logic needed updating
> 
> I see what you're saying. I think the best we could do in this case would be to write a test suite to represent what we expect to happen for certain inputs. That way, we should know if something breaks that expectation without having to manually run different variations and compare debug output.
> 

I'll take the debug code out. Unit tests probably are the way to go. 
Long term, having some kind of visualisation tool like space finder might be worthwhile.

[the remark this answers is not in the document]
Comment (dblatcher) [reply apps-rendering/src/server/insertNewsletterIntoItem.ts:92]: I think we should still document the adslot rule in the comment as it is one of the requirements for the sign-up blocks. I've updated the line to explain how it is currently satisfied by the adslot placement.

File: apps-rendering/src/server/insertNewsletterIntoItem.test.ts [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+const makeTextElementNode = (html: string, outerTag = 'p'): Node => {
+	const doc = new DocumentFragment();
+	const el = document.createElement(outerTag);
+	el.innerHTML = html;
+	doc.appendChild(el);
+	// eslint-disable-next-line @typescript-eslint/no-non-null-assertion -- this value is not `null`
+	return doc.firstChild!;
+};
+
+const makeBodyWithPlaceToInsert: { (): Body } = () => [
```
[the remark this answers is not in the document]
Comment (dblatcher) [reply apps-rendering/src/server/insertNewsletterIntoItem.test.ts:34]: looking at this again, the function type isn't really needed - there are no parameters, so I should just be declaring the type on the return value: 
```
const makeBodyWithPlaceToInsert = (): Body => [
```

File: apps-rendering/src/server/insertNewsletterIntoItem.test.ts [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+	Result.ok({
+		kind: ElementKind.Text,
+		doc: makeTextElementNode('Introductory paragraph.'),
+	}),
```
[the remark this answers is not in the document]
Comment (dblatcher) [reply apps-rendering/src/server/insertNewsletterIntoItem.test.ts:38]: I think they should be unique(as in, non-identical) as having the same constant repeated in the array doesn't replicate an actual `body`, where each block is a different object. 

Having the same constant could potentially lead to broken tests if I forgot to account for it e.g.: 
```
list = [Paragraph, Paragraph, Paragraph]
secondMember = list[1]
expect(list.indexOf(secondMember)).toBe(1) // actually 0
```

However, we could make it a lot cleaner with by using a paragraph function, so the members of the list are different objects (but with the same value), eg:
```
return [
 makeParagraph(),
 makeParagraph(),
 makeParagraph()
]

```

File: apps-rendering/src/server/insertNewsletterIntoItem.ts [hunk 13]
```diff
@@ -0,0 +1,264 @@
...
+ */
+function getRange(numberOfElements: number): [number, number] {
+	const middle = Math.floor(numberOfElements / 2);
+	return [Math.max(0, middle - 3), Math.min(numberOfElements, middle + 3)];
+}
+
+/**
+ *  Find the place to insert a insert a sign-up block in to an article body,
+ *  using these rules:
+ *	 - Don't put straight after bold text
```
[the remark this answers is not in the document]
Comment (dblatcher) [reply apps-rendering/src/server/insertNewsletterIntoItem.ts:83]: not quite - `contentBefore` is the element that would be **before the signup block** (so the one the signup block , will go after. 
Now I read that back I can see that's confusing! `contentBefore` could be a name for the content that the signUp will be before or a name for  the content that will be before the signup

So, to look at the whole conditional:
```
			[ElementCategory.ParagraphText].includes(
				categoriseElement(contentBefore),
			) &&
			[
				ElementCategory.ParagraphText,
				ElementCategory.BoldParagraphText,
			].includes(categoriseElement(contentAfter))
```
the place is only suitable if it would go like this:

contentBefore: ParagraphText
[signup goes here]
contentAfter: ParagraphText or BoldParagraphText

[the remark this answers is not in the document]
Comment (georgeblahblah) [reply apps-rendering/src/server/insertNewsletterIntoItem.ts:83]: Thank you for the explanation! That's really helpful. I think I was reading the logic in the wrong direction! The way it is now makes sense

File: apps-rendering/src/server/insertNewsletterIntoItem.test.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (dblatcher) [reply apps-rendering/src/server/insertNewsletterIntoItem.test.ts]: I've made that change - it is a lot easier to see the structure now there's one line for each element in the body.

File: apps-rendering/src/newsletter.ts [hunk 14]
```diff
@@ -0,0 +1,254 @@
...
+	}
+
+	// Define a recursive function go backwards through the possible
+	// positions until it finds one that is suitable or reaches
+	// possibleElementsToPlaceBefore[0] - which is actually out of range, but
+	// has to be include to test if possiblePoistions[1] is suitable.
+	const findLastSuitablePosition = (
+		idx: number,
+	): Result<string, ElementCategoryAndIndex> => {
+		if (idx <= 0) {
```
Comment (georgeblahblah) [inline apps-rendering/src/newsletter.ts:157]: I think this could end up searching all the way back to the beginning of the body, which might be a bug?

**Review by georgeblahblah — APPROVED** [review]
Looking good! I had one comment about hunting for a possible position back to index `0` which I think could be a source of bugs.

Otherwise, this feature is really exciting. Thank you so much for your patience!

File: apps-rendering/src/newsletter.ts (hunk 14 above)
Comment (georgeblahblah) [reply apps-rendering/src/newsletter.ts]: That makes sense! I hadn't realised the array was sliced beforehand

Comment (dblatcher) [reply apps-rendering/src/newsletter.ts]: `idx` is the index within `possibleElementsToPlaceBefore`, the array of `ElementCategoryAndIndex` , which is already bound by `minIndex` and `maxIndex` - (IE it's `possibleElementsToPlaceBefore` that is being iterated over backward, not the whole body).

I've rename the variable to be clearer which index we are using.

Also added a new test to make sure the signup block can only be placed in the middle.
